api/pkg/transformer/pipeline/operation.go

```diff
@@ -1,7 +1,118 @@
 package pipeline
 
-import "context"
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/gojek/merlin/pkg/transformer/spec"
+	"github.com/gojek/merlin/pkg/transformer/types"
+	"github.com/gojek/merlin/pkg/transformer/types/table"
+)
 
 type Op interface {
 	Execute(context context.Context, environment *Environment) error
+	AddInputOutput(input, output map[string]interface{}) error
+	GetOperationTracingDetail() ([]types.TracingDetail, error)
+}
+
+type OperationIO struct {
+	Name  string
+	Value interface{}
+}
+
+type OperationTracing struct {
+	mutex  sync.RWMutex
+	Input  []map[string]interface{}
+	Output []map[string]interface{}
+	Specs  interface{}
+	OpType types.OperationType
+}
+
+func NewOperationTracing(operationSpecs interface{}, opType types.OperationType) *OperationTracing {
+	opTracing := &OperationTracing{
+		Specs:  operationSpecs,
+		OpType: opType,
+	}
+	return opTracing
+}
+
+func (ot *OperationTracing) AddInputOutput(input, output map[string]interface{}) error {
+	ot.mutex.Lock()
+	defer ot.mutex.Unlock()
+
+	if ot.Input == nil {
+		ot.Input = make([]map[string]interface{}, 0)
+	}
+	if ot.Output == nil {
+		ot.Output = make([]map[string]interface{}, 0)
+	}
+
+	if err := sanitizeIO(input); err != nil {
+		return err
+	}
+	if err := sanitizeIO(output); err != nil {
+		return err
+	}
+
+	ot.Input = append(ot.Input, input)
+	ot.Output = append(ot.Output, output)
+
+	return nil
+}
+
+func sanitizeIO(io map[string]interface{}) error {
```

Comment:
In what case `sanitizeIO` is required?

Reply:
it is to change `table` value to formatted value because `table` can't be marshalled without create custom marshaller